frontend/ui/widget/textboxwidget.lua

```diff
@@ -2186,4 +2176,247 @@ function TextBoxWidget:getSourceIndex(char_idx)
+    -- So, just find out charpos (text indice) of both points and
+    -- find word edges in the logical order text/charlist.
+    local sel_start_idx = self:getCharPosAtXY(start_x, start_y)
+    local sel_end_idx = self:getCharPosAtXY(end_x, end_y)
+    if not sel_start_idx or not sel_end_idx then
+        -- one or both hold points were out of text
+        return nil, nil, nil, nil
+    end
+    if sel_start_idx > sel_end_idx then -- re-order if needed
+        sel_start_idx, sel_end_idx = sel_end_idx, sel_start_idx
+    end
+    -- We get cursor positions, which can be after last char,
+    -- and that we need to correct. But if both positions are
+    -- after last char, the full selection is out of text.
+    if sel_start_idx > #self._xtext then -- Both are after last char
+        return nil, nil, nil, nil
+    end
+    if sel_end_idx > #self._xtext then -- Only end is after last char
+        sel_end_idx = #self._xtext
+    end
+
+    local start_line_num = math.ceil(start_y / self.line_height_px) + self.virtual_line_num - 1
+    local end_line_num = math.ceil(end_y / self.line_height_px) + self.virtual_line_num - 1
+    if start_line_num < 1 or end_line_num > #self.vertical_string_list then
+        return nil, nil, nil, nil
+    end
+
+    local word_start_idx, word_end_idx = self._xtext:getSelectedWordIndices(sel_start_idx, sel_end_idx, 50)
+    return word_start_idx, word_end_idx, start_line_num, end_line_num
+end
+
+function TextBoxWidget:getXtextHighlightRects(text_start_idx, text_end_idx, start_line_num, end_line_num)
+    local rects = {}
+
+    for line_num = start_line_num, end_line_num, 1 do
+        local line = self.vertical_string_list[line_num]
+        local draw_line = false
+        local line_x0 = 0
+        local line_x1 = 0
+
+        for _, xglyph in ipairs(line.xglyphs) do
+            if xglyph.text_index >= text_start_idx and (xglyph.text_index + xglyph.cluster_len) <= text_end_idx and (not xglyph.no_drawing) then
+                if draw_line then
+                    if xglyph.x0 < line_x0 then line_x0 = xglyph.x0 end
+                    if xglyph.x1 > line_x1 then line_x1 = xglyph.x1 end
+                else
+                    draw_line = true
+                    line_x0 = xglyph.x0
+                    line_x1 = xglyph.x1
+                end
+            end
+        end
+
+        if draw_line then
+            local rect = Geom:new{
+                x = line_x0,
+                y = (line_num - self.virtual_line_num) * self.line_height_px,
+                w = line_x1 - line_x0,
+                h = self.line_height_px,
+            }
+
+            table.insert(rects, rect)
+        end
+    end
+
+    return rects
+end
+
+function TextBoxWidget:getNonXtextHighlightIndices(start_x, start_y, end_x, end_y)
+    local start_idx = self:_findWordEdge(start_x, start_y, FIND_START)
+    local end_idx = self:_findWordEdge(end_x, end_y, FIND_END)
+    if (not start_idx) or (not end_idx) then
```

Comment:
(We usually don't (or don't need to) wrap these `not` with parens, dunno if these are really needed ?)